src/box/engine.h

```diff
@@ -315,11 +316,27 @@ struct engine_read_view {
 	struct rlist link;
 };
 
+/**
+ * Cursor used during checkpoint initial join. Shared between engines.
+ */
+struct checkpoint_cursor {
+	/** Signature of the checkpoint to take data from. */
+	struct vclock *vclock;
+	/** Checkpoint lsn to start from. */
+	int64_t start_lsn;
+	/** Counter, shared between engines */
+	int64_t lsn_counter;
+};
+
 struct engine_join_ctx {
 	/** Vclock to respond with. */
 	struct vclock *vclock;
 	/** Whether sending JOIN_META stage is required. */
 	bool send_meta;
+	/** Checkpoint join cursor. */
+	struct checkpoint_cursor *cursor;
+	/** Ref to protect checkpoint from gc during checkpoint join. */
+	struct gc_checkpoint_ref gc;
```

Comment:
I don't like that `engine_join_ctx::gc` is defined in CE but used only in EE. Let's store it in a context allocated in EE.

Reply:
Good idea. The code is cleaner now